First commits

jsf9k opened the pull request:
## 🗣 Description ##

This pull request implements the Python code for an AWS Lambda function that will disable users' inactive console and programmatic access.

See also:
- cisagov/provisionaccount-role-tf-module#45
- cisagov/disable-inactive-iam-users-tf-module#1
- cisagov/cool-accounts#147

## 💭 Motivation and context ##

We have a POAM about disabling inactive users in IAM.  Partially resolves cisagov/cool-system-internal#143.

## 🧪 Testing ##

I applied this module to our COOL Users account without incident.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [x] Finalize version.

## ✅ Post-merge checklist ##

- [x] Create a release.

jsf9k committed:
First cut at creating this Lambda function

jsf9k committed:
Ensure that expiration_days is nonzero

jsf9k committed:
Correct the way I was using boto3 paginators

jsf9k committed:
Ignore access keys that are already inactive

jsf9k committed:
Handle the case where a user's console access was created recently

If a user's console access was created too recently then we cannot yet
determine inactivity, so we should not disable it.

jsf9k committed:
Handle the case where a user's access key was created recently

If a user's access key was created too recently then we cannot yet
determine inactivity, so we should not disable it.

jsf9k committed:
Update Pipfiles (and lock files) for this project

jsf9k committed:
Update Docker composition to match the needs of this project

I also removed the run_lambda_locally section as it will not be used
with this project.

jsf9k committed:
Correctly use boto3 IAM client exception class

jsf9k committed:
Make now a timezone-aware datetime object

Otherwise we cannot perform arithmetic such as now -
login_profile_create, since the latter is timezone-aware.

jsf9k committed:
Remove boto3 from the Pipfiles

This package is already available to Lambda functions.

jsf9k committed:
Correctly extract the last used date from the get_access_key_last_used response

jsf9k committed:
Correctly handle the case where the user has never used his or her console access

In such a case the "PasswordLastUsed" key does not exist for the user.

jsf9k committed:
Add a debug message for the case where an access key is already inactive

jsf9k committed:
Word-wrap paragraph to get rid of linting error

jsf9k committed:
Correct reference to repo that does not exist

jsf9k committed:
Uncomment code that actually does the disabling

jsf9k committed:
Bump version from 0.0.2 to 1.0.0

jsf9k committed:
Bump version from 1.0.0 to 1.0.0-rc.1

jsf9k committed:
Set author label in Dockerfile to the VM dev distro

dav3r commented on src/lambda_handler.py:
```diff
@@ -36,50 +40,159 @@ def task_default(event):
+    # integer.
+    if "expiration_days" not in event:
+        errors.append('Missing required key "expiration_days" in event.')
+    elif not event["expiration_days"]:
+        errors.append('"account_ids" must be non-null.')
```
Looks like you copy/pasted this from elsewhere, but didn't fully update it.  Is this what you meant?
```suggestion
    elif not event["expiration_days"]:
        errors.append('"expiration_days" must be non-null.')
```

dav3r commented on src/lambda_handler.py:
```diff
@@ -36,50 +40,159 @@ def task_default(event):
-            logging.error(variable_error_msg, numerator, denominator)
-            failed_task(result, error_msg)
-
+            errors.append('"account_ids" must be an integer.')
```
Fix copy/paste error:
```suggestion
            errors.append('"expiration_days" must be an integer.')
```

dav3r commented on src/lambda_handler.py:
```diff
@@ -36,50 +40,159 @@ def task_default(event):
-            result["result"] = example_div(int(numerator), int(denominator))
+            tmp = int(event["expiration_days"])
+            if tmp <= 0:
+                errors.append('"account_ids" must be a strictly positive integer.')
```
Fix copy/paste error:
```suggestion
                errors.append('"expiration_days" must be a strictly positive integer.')
```

dav3r reviewed (approved):
Looks great!  I noted a few thangs in `validate_event_data` that I think you overlooked.

jsf9k committed:
Fix copy-and-paste errors

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj reviewed (changes requested):
Please enable the dependabot ignore directives and fix the upstream repository reference in https://github.com/cisagov/disable-inactive-iam-users-lambda/blob/15fc27e20945f2a7a7945535c4a54647a6c1ce1f/.github/dependabot.yml#L22-L24

That reference changing suggests you have not updated your locally installed `skeleton` extension. Please update with `gh extension upgrade --all` to ensure you're working with the latest copy.

jsf9k committed:
Uncomment Dependabot ignore lines

Also correct repo reference in comment.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented:
> Please enable the dependabot ignore directives and fix the upstream repository reference in
> 
> https://github.com/cisagov/disable-inactive-iam-users-lambda/blob/15fc27e20945f2a7a7945535c4a54647a6c1ce1f/.github/dependabot.yml#L22-L24
> 
> That reference changing suggests you have not updated your locally installed `skeleton` extension. Please update with `gh extension upgrade --all` to ensure you're working with the latest copy.

Please see commit 469a51f.

jsf9k committed:
Simplify timestamp logic

dv4harr10 commented:
Hi Team, for README.md file section 'How to update Python dependencies' sentence 'More information about the Pipfile format can be found here.' The link from 'here' returns a 404 not found error.

jsf9k committed:
Add pydoc for task_disable() parameters

jsf9k commented:
> Hi Team @ALL , for README.md file section 'How to update Python dependencies' sentence 'More information about the Pipfile format can be found here.' The link from 'here' returns a 404 not found error.

I created cisagov/skeleton-aws-lambda-python#32 to fix this upstream.

jsf9k committed:
Bump version from 1.0.0-rc.1 to 1.0.0-rc.2

jsf9k committed:
Add support for Python 3.10 and remove support for older versions of Python

This is the latest Python runtime we can currently support, given our
AWS Terraform provider constraints.

Co-authored-by: Dave Redmin <david.redmin@gwe.cisa.dhs.gov>

jsf9k committed:
Update workflow to only support Python 3.10 Lambda runtime

jsf9k committed:
Finalize version from 1.0.0-rc.2 to 1.0.0

jsf9k commented:
@mcdonnnj - This work is due tomorrow, and you are out of the office, so I am going to assume that you are satisfied with the changes I made in response to your review.
